src/main/java/com/diffblue/deeptestutils/mock/DTUMemberMatcher.java

```diff
@@ -0,0 +1,155 @@
+
+    try {
+      return org.powermock.api.support.membermodification.MemberMatcher
+          .method(declaringClass, methodName, parameterTypes);
+    } catch (TooManyMethodsFoundException e) {
+      return methodUsingWholeSignature(declaringClass, methodName,
+          parameterTypes);
+    }
+  }
+
+  /**
+   * Get a method matching the given name and parameter types. If there
+   * are multiple methods then pick the one that would be chosen by Java at
+   * runtime.
+   * <p>
+   * We first gather all methods from the class and its ancestors that can
+   * be considered as a match and then find the best match using
+   * MethodComparatorFull.
+   * <p>
+   * Note that this does not handle generics, variadic arguments, arrays and
+   * possibly other cases correctly. For full spec on method invocation see
+   * https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.12
+   * <p>
+   * This is a helper method that is not intended to be called by users,
+   * instead use the "method" method.
+   *
+   * @param declaringClass The class to search
+   * @param methodName     The name of the desired method
+   * @param parameterTypes Types that can be fed as arguments to the method
+   *                       (can be more specific than the method's declared
+   *                       types)
+   * @return A matching method if such exists
+   */
+  protected static Method methodUsingWholeSignature(
+      final Class<?> declaringClass,
+      final String methodName,
+      final Class<?>... parameterTypes) {
+
+    Method[] declaredMethods = WhiteboxImpl.getMethods(
+        declaringClass, methodName, parameterTypes, false);
+    Arrays.sort(declaredMethods, new MethodComparatorFull());
+
+    if (declaredMethods.length == 0) {
+      WhiteboxImpl.throwExceptionIfMethodWasNotFound(declaringClass,
+          methodName, null, new Object[]{parameterTypes});
+      return null;
+    } else {
+      return declaredMethods[0];
+    }
+  }
+
+}
+
+/**
+ * Comparator for methods that takes into consideration the arguments,
+ * declaring class and return type, in this order.
+ */
+class MethodComparatorFull implements Comparator<Method> {
+  /**
+   * Comparator for arguments (i.e. ignores return types and declaring
+   * classes)
+   */
+  private Comparator<Method> argOnlyComparator =
+      ComparatorFactory.createMethodComparator();
+
+  @Override
+  public int compare(final Method m1, final Method m2) {
+    // first try to decide order based on arguments
+    int argCompare = argOnlyComparator.compare(m1, m2);
+
+    if (argCompare != 0) {
+      return argCompare;
+    }
+
+    // if arguments are the same, then try to pick based on the
+    // declaring class
+    Class<?> declType1 = m1.getDeclaringClass();
+    Class<?> declType2 = m2.getDeclaringClass();
+
+    if (declType1.equals(declType2)) {
+      // if declaring class is the same, pick the most special return type
```

Comment:
:question:  Where is the source of truth for that behaviour? Is that specified in the JLS? Might be worth adding a reference here.